File: tape/sdk/python/tape/chaos/invariants.py
"""Invariants — predicates over Tape's journal projections.

The journal *is* the oracle. Every invariant here reads from
projections that already exist in Tape (the WAL, `tape_effects`,
`tape_obligations`, the budget state, …); none of them require a parallel
test ledger.

The catalogue:

  exactly_one(connector, by="business_key")  — duplicate-effect detector
  no_stuck_obligations                       — every obligation drains
  no_budget_overrun                          — spend stays within cap
  no_blind_non_idempotent_retry              — the §6.5 safety property
  no_orphan_compensation                     — every obligation has an effect
  no_effect_without_decision                 — every effect has its parent

Each invariant has:

  * `name`            — for the report
  * `check(client, run_id) -> InvariantResult` — the predicate

Per-run invariants (most) want `run_id`. Cross-run ones (`exactly_one`)
ignore it. The runner threads it through; tests just pass the run id they
created.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from ..client import (
    TapeClient,
    EFFECT_STATUS_CONFIRMED,
    EFFECT_STATUS_PENDING,
    EFFECT_STATUS_UNKNOWN,
    EFFECT_SEMANTICS_NON_IDEMPOTENT,
    OBLIGATION_STATUS_STUCK,
    OBLIGATION_STATUS_PENDING,
    OBLIGATION_STATUS_COMMITTED,
)


@dataclass
class InvariantResult:
    """The outcome of one invariant check."""
    name: str
    passed: bool
    detail: str = ""

    def __str__(self) -> str:
        mark = "OK " if self.passed else "FAIL"
        return f"[{mark}] {self.name}: {self.detail}" if self.detail else f"[{mark}] {self.name}"


# ── Base ────────────────────────────────────────────────────────────────────

class Invariant:
    """A predicate over the journal. Subclasses override `name` and `check`."""
    name: str = "<unnamed>"

    def check(self, *, client: TapeClient, run_id: Optional[str]) -> InvariantResult:
        raise NotImplementedError


# ── Concrete invariants ─────────────────────────────────────────────────────

class _ExactlyOne(Invariant):
    """For all CONFIRMED effects under one connector that share a business
    key, count = 1. This is the "one wire, one record" property the treasury
    test claims.

    Implementation: walks the cross-run journal tail via `SubscribeEvents`
    with a subject pattern on `/tape/effect/confirmed/**`, then groups by
    `business_key` (when set). Effects without a business key are skipped —
    they can't be deduped at the business level and so don't admit this
    invariant.
    """

    def __init__(self, *, connector: str = "", tool: str = "",
                 by: str = "business_key"):
        if not connector and not tool:
            raise ValueError("exactly_one needs connector= or tool=")
        self.connector = connector
        self.tool = tool
        self.by = by
        self.name = f"exactly_one({connector or tool!r}, by={by!r})"

    def check(self, *, client: TapeClient, run_id: Optional[str]) -> InvariantResult:
        # v1: walk the journal via SubscribeEvents with subject prefix. The
        # subject for a confirmed effect is `/tape/effect/confirmed/<tool>/<run_id>`;
        # we filter post-hoc by the connector tag stored in the payload.
        from .._gen import tape_pb2 as pb
        pattern = "/tape/effect/confirmed/**"
        if self.tool:
            pattern = f"/tape/effect/confirmed/{self.tool}/**"

        counts: dict[str, int] = {}
        try:
            it = client.subscribe_events(from_global_seq=1, subject_pattern=pattern)
        except Exception as ex:
            return InvariantResult(self.name, passed=False,
                                    detail=f"subscribe_events failed: {ex}")
        # The stream is long-lived; iterate non-blockingly by draining with a
        # short deadline.
        import json
        import time
        deadline = time.time() + 2.0
        try:
            for evt in it:
                try:
                    payload = json.loads(evt.payload_json or "{}")
                except Exception:
                    continue
                connector = str(payload.get("connector") or "")
                bk = str(payload.get(self.by) or "")
                if self.connector and connector != self.connector:
                    continue
                if not bk:
                    continue
                counts[bk] = counts.get(bk, 0) + 1
                if time.time() > deadline:
                    break
        finally:
            try:
                it.cancel()  # type: ignore[attr-defined]
            except Exception:
                pass

        dupes = {k: v for k, v in counts.items() if v > 1}
        if dupes:
            return InvariantResult(self.name, passed=False,
                                    detail=f"duplicate business keys: {dupes}")
        return InvariantResult(self.name, passed=True,
                                detail=f"unique business keys: {len(counts)}")


def exactly_one(*, connector: str = "", tool: str = "",
                by: str = "business_key") -> Invariant:
    return _ExactlyOne(connector=connector, tool=tool, by=by)


class _NoStuckObligations(Invariant):
    name = "no_stuck_obligations"

    def check(self, *, client: TapeClient, run_id: Optional[str]) -> InvariantResult:
        try:
            resp = client.list_unresolved_obligations(
                include_stuck=True, include_pending=False,
                include_committed_expired=False, limit=500)
        except Exception as ex:
            return InvariantResult(self.name, passed=False,
                                    detail=f"list_unresolved_obligations failed: {ex}")
        stuck = [o for o in resp.obligations
                 if o.status == OBLIGATION_STATUS_STUCK
                 and (not run_id or o.run_id == run_id)]
        if stuck:
            return InvariantResult(self.name, passed=False,
                                    detail=f"{len(stuck)} stuck obligation(s)")
        return InvariantResult(self.name, passed=True, detail="0 stuck")


no_stuck_obligations: Invariant = _NoStuckObligations()


class _NoBlindNonIdempotentRetry(Invariant):
    """For every NON_IDEMPOTENT effect: either there is exactly one dispatch
    attempt, or the reconciler has recorded an external observation (the
    effect carries an external_ref or has been resolved). The unsafe case
    is `dispatch_attempts > 1 AND status == PENDING AND external_ref == ""`
    — the outbox reactor re-dispatched a non-idempotent effect without
    observing first, which is exactly the property §6.5 forbids."""

    name = "no_blind_non_idempotent_retry"

    def check(self, *, client: TapeClient, run_id: Optional[str]) -> InvariantResult:
        try:
            resp = client.list_pending_effects(
                include_pending=True, include_unknown=True, limit=500)
        except Exception as ex:
            return InvariantResult(self.name, passed=False,
                                    detail=f"list_pending_effects failed: {ex}")
        bad = []
        for e in resp.effects:
            if run_id and e.run_id != run_id:
                continue
            if e.semantics != EFFECT_SEMANTICS_NON_IDEMPOTENT:
                continue
            if e.dispatch_attempts > 1 and e.status == EFFECT_STATUS_PENDING and not e.external_ref:
                bad.append((e.run_id, e.idempotency_key, e.dispatch_attempts))
        if bad:
            return InvariantResult(self.name, passed=False,
                                    detail=f"{len(bad)} non-idempotent effect(s) re-dispatched without observation: {bad[:3]}")
        return InvariantResult(self.name, passed=True,
                                detail="no blind retries on non-idempotent effects")


no_blind_non_idempotent_retry: Invariant = _NoBlindNonIdempotentRetry()


class _NoBudgetOverrun(Invariant):
    """`get_run(run_id)` carries the budget state. Spent must not exceed cap."""

    name = "no_budget_overrun"

    def check(self, *, client: TapeClient, run_id: Optional[str]) -> InvariantResult:
        if not run_id:
            return InvariantResult(self.name, passed=True,
                                    detail="no run_id; skipped (cross-run budget check is Phase 3)")
        try:
            run = client.get_run(run_id)
        except Exception as ex:
            return InvariantResult(self.name, passed=False,
                                    detail=f"get_run failed: {ex}")
        # The budget is a separate projection; query via the proto field if
        # present, otherwise fall back to a per-run query. v1: stub —
        # the run record doesn't carry budget directly; the test asserts on
        # `client.charge_budget` returns. Phase 3 adds GetBudget to the proto.
        return InvariantResult(self.name, passed=True,
                                detail="budget projection check is a Phase-3 invariant; v1 stub")


no_budget_overrun: Invariant = _NoBudgetOverrun()


class _NoOrphanCompensation(Invariant):
    """Every obligation must reference an existing effect."""

    name = "no_orphan_compensation"

    def check(self, *, client: TapeClient, run_id: Optional[str]) -> InvariantResult:
        if not run_id:
            return InvariantResult(self.name, passed=True,
                                    detail="no run_id; skipped (cross-run scan is Phase 3)")
        try:
            obs = client.list_obligations(run_id=run_id).obligations
        except Exception as ex:
            return InvariantResult(self.name, passed=False,
                                    detail=f"list_obligations failed: {ex}")
        orphans = []
        for o in obs:
            try:
                got = client.get_effect(run_id=run_id, idempotency_key=o.effect_key)
                if not got.found:
                    orphans.append(o.effect_key)
            except Exception:
                orphans.append(o.effect_key)
        if orphans:
            return InvariantResult(self.name, passed=False,
                                    detail=f"{len(orphans)} obligation(s) with no effect: {orphans[:3]}")
        return InvariantResult(self.name, passed=True,
                                detail=f"all {len(obs)} obligation(s) have an effect")


no_orphan_compensation: Invariant = _NoOrphanCompensation()


__all__ = [
    "Invariant",
    "InvariantResult",
    "exactly_one",
    "no_stuck_obligations",
    "no_blind_non_idempotent_retry",
    "no_budget_overrun",
    "no_orphan_compensation",
]

File: tape/sdk/java/src/main/java/dev/tape/chaos/Invariants.java
package dev.tape.chaos;

import dev.tape.TapeClient;
import dev.tape.proto.EffectStatus;
import dev.tape.proto.EffectSemantics;
import dev.tape.proto.EventEntry;
import dev.tape.proto.ListPendingEffectsResponse;
import dev.tape.proto.ListUnresolvedObligationsResponse;
import dev.tape.proto.ObligationRecord;
import dev.tape.proto.ObligationStatus;
import dev.tape.proto.EffectRecord;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * Built-in invariants — predicates over Tape's journal projections.
 * Mirrors {@code tape.chaos.invariants} (Python/TS/Go).
 */
public final class Invariants {
    private Invariants() {}

    // ── exactly_one ──────────────────────────────────────────────────────

    /** "One wire, one record" — every CONFIRMED effect with a non-empty
     *  business_key has count == 1 under {@code connector} (or {@code tool}). */
    public static Invariant exactlyOne(String connector, String tool, String by) {
        if ((connector == null || connector.isEmpty()) && (tool == null || tool.isEmpty())) {
            throw new IllegalArgumentException("exactlyOne: needs connector or tool");
        }
        String byField = (by == null || by.isEmpty()) ? "business_key" : by;
        String target = (connector != null && !connector.isEmpty()) ? connector : tool;
        String name = "exactly_one(\"" + target + "\", by=\"" + byField + "\")";

        return new Invariant() {
            @Override public String name() { return name; }
            @Override public InvariantResult check(TapeClient client, String runId) {
                String pattern = (tool != null && !tool.isEmpty())
                        ? "/tape/effect/confirmed/" + tool + "/**"
                        : "/tape/effect/confirmed/**";
                Map<String, Integer> counts = new HashMap<>();
                try {
                    Iterator<EventEntry> it = client.subscribeBySubject(pattern, "", 1);
                    while (it.hasNext()) {
                        EventEntry e = it.next();
                        Map<String, Object> payload = SimpleJson.parseObject(e.getPayloadJson());
                        if (payload == null) continue;
                        if (connector != null && !connector.isEmpty()
                            && !connector.equals(String.valueOf(payload.get("connector")))) continue;
                        Object k = payload.get(byField);
                        if (!(k instanceof String) || ((String) k).isEmpty()) continue;
                        counts.merge((String) k, 1, Integer::sum);
                    }
                } catch (Exception ex) {
                    return InvariantResult.fail(name, "subscribeBySubject failed: " + ex.getMessage());
                }
                StringBuilder dupes = new StringBuilder();
                int dupeCount = 0;
                for (Map.Entry<String, Integer> e : counts.entrySet()) {
                    if (e.getValue() > 1) {
                        if (dupeCount > 0) dupes.append(",");
                        dupes.append(e.getKey()).append(":").append(e.getValue());
                        dupeCount++;
                    }
                }
                if (dupeCount > 0) {
                    return InvariantResult.fail(name, "duplicate business keys: " + dupes);
                }
                return InvariantResult.ok(name, "unique business keys: " + counts.size());
            }
        };
    }

    // ── no_stuck_obligations ─────────────────────────────────────────────

    public static final Invariant NO_STUCK_OBLIGATIONS = new Invariant() {
        @Override public String name() { return "no_stuck_obligations"; }
        @Override public InvariantResult check(TapeClient client, String runId) {
            try {
                ListUnresolvedObligationsResponse resp = client.listUnresolvedObligations(
                        500, 0L, /*includePending*/ false, /*includeStuck*/ true,
                        /*includeCommittedExpired*/ false);
                int stuck = 0;
                for (ObligationRecord o : resp.getObligationsList()) {
                    if (o.getStatus() != ObligationStatus.OBLIGATION_STATUS_STUCK) continue;
                    if (runId != null && !runId.isEmpty() && !runId.equals(o.getRunId())) continue;
                    stuck++;
                }
                if (stuck > 0) return InvariantResult.fail(name(), stuck + " stuck obligation(s)");
                return InvariantResult.ok(name(), "0 stuck");
            } catch (Exception ex) {
                return InvariantResult.fail(name(), "listUnresolvedObligations failed: " + ex.getMessage());
            }
        }
    };

    // ── no_blind_non_idempotent_retry ────────────────────────────────────

    public static final Invariant NO_BLIND_NON_IDEMPOTENT_RETRY = new Invariant() {
        @Override public String name() { return "no_blind_non_idempotent_retry"; }
        @Override public InvariantResult check(TapeClient client, String runId) {
            try {
                ListPendingEffectsResponse resp = client.listPendingEffects(
                        0L, /*includePending*/ true, /*includeUnknown*/ true, 500L);
                int bad = 0;
                StringBuilder head = new StringBuilder();
                int headN = 0;
                for (EffectRecord e : resp.getEffectsList()) {
                    if (runId != null && !runId.isEmpty() && !runId.equals(e.getRunId())) continue;
                    if (e.getSemantics() != EffectSemantics.EFFECT_SEMANTICS_NON_IDEMPOTENT) continue;
                    if (e.getDispatchAttempts() > 1
                            && e.getStatus() == EffectStatus.EFFECT_STATUS_PENDING
                            && e.getExternalRef().isEmpty()) {
                        bad++;
                        if (headN < 3) {
                            if (headN > 0) head.append(", ");
                            head.append(e.getRunId()).append("/").append(e.getIdempotencyKey())
                                .append("@").append(e.getDispatchAttempts());
                            headN++;
                        }
                    }
                }
                if (bad > 0) {
                    return InvariantResult.fail(name(),
                            bad + " non-idempotent effect(s) re-dispatched without observation: " + head);
                }
                return InvariantResult.ok(name(), "no blind retries on non-idempotent effects");
            } catch (Exception ex) {
                return InvariantResult.fail(name(), "listPendingEffects failed: " + ex.getMessage());
            }
        }
    };

    // ── no_orphan_compensation ───────────────────────────────────────────

    public static final Invariant NO_ORPHAN_COMPENSATION = new Invariant() {
        @Override public String name() { return "no_orphan_compensation"; }
        @Override public InvariantResult check(TapeClient client, String runId) {
            if (runId == null || runId.isEmpty()) return InvariantResult.ok(name(), "no runId; skipped");
            try {
                var resp = client.listObligations(runId, false);
                int orphans = 0;
                StringBuilder head = new StringBuilder();
                int headN = 0;
                for (ObligationRecord o : resp.getObligationsList()) {
                    var got = client.getEffect(runId, o.getEffectKey());
                    if (!got.getFound()) {
                        orphans++;
                        if (headN < 3) {
                            if (headN > 0) head.append(", ");
                            head.append(o.getEffectKey());
                            headN++;
                        }
                    }
                }
                if (orphans > 0) {
                    return InvariantResult.fail(name(),
                            orphans + " obligation(s) with no effect: " + head);
                }
                return InvariantResult.ok(name(),
                        "all " + resp.getObligationsCount() + " obligation(s) have an effect");
            } catch (Exception ex) {
                return InvariantResult.fail(name(), "listObligations failed: " + ex.getMessage());
            }
        }
    };

    // ── no_budget_overrun (v1 stub) ──────────────────────────────────────

    public static final Invariant NO_BUDGET_OVERRUN = new Invariant() {
        @Override public String name() { return "no_budget_overrun"; }
        @Override public InvariantResult check(TapeClient client, String runId) {
            if (runId == null || runId.isEmpty()) return InvariantResult.ok(name(), "no runId; skipped");
            return InvariantResult.ok(name(), "budget projection check is a Phase-3 invariant; v1 stub");
        }
    };
}
